File: Blind 75/16 - Intervals/435-non-overlapping-intervals.py
'''
Given an array of intervals intervals where intervals[i] = [start_i, end_i], return the minimum number of intervals you need to remove to make the rest of the intervals non-overlapping.
'''

from time import time
from typing import List


class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x: x[1])
        prev = 0
        valid_intervals = 1

        for i in range(1, len(intervals)):
            # Check how many intervals are valid
            if intervals[prev][1] <= intervals[i][0]:
                prev = i
                valid_intervals += 1

        # Invalid intervals is the difference
        return len(intervals) - valid_intervals

    def reference(self, intervals: List[List[int]]) -> int:
        intervals.sort()
        res = 0
        prevEnd = intervals[0][1]
        for start, end in intervals[1:]:
            if start >= prevEnd:
                prevEnd = end
            else:
                res += 1
                prevEnd = min(end, prevEnd)
        return res

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.eraseOverlapIntervals(case))
                else:
                    self.eraseOverlapIntervals(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        [[1, 2], [2, 3], [3, 4], [1, 3]],
        [[1, 2], [1, 2], [1, 2]],
        [[1, 2], [2, 3]],
        # Additional
        [
            [-52, 31],
            [-73, -26],
            [82, 97],
            [-65, -11],
            [-62, -49],
            [95, 99],
            [58, 95],
            [-31, 49],
            [66, 98],
            [-63, 2],
            [30, 47],
            [-40, -26],
        ],
        [
            [-52, 31],
            [-73, -26],
            [82, 97],
            [-65, -11],
            [-62, -49],
            [95, 99],
            [58, 95],
            [-31, 49],
            [66, 98],
            [-63, 2],
            [30, 47],
            [-40, -26],
        ],
    ]
    test.quantify(test_cases)

File: Blind 75/16 - Intervals/JavaScript/435-non-overlapping-intervals.js
const { performance } = require('perf_hooks');

class Solution {
  method(intervals) {
    let count = 0;
    let prevEnd = -Infinity;
    intervals.sort((a, b) => a[0] - b[0]);
    for (const [start, end] of intervals) {
      if (prevEnd && prevEnd > start) {
        count += 1;
        prevEnd = Math.min(prevEnd, end);
      } else {
        prevEnd = end;
      }
    }

    return count;
  }

  reference(intervals) {
    intervals.sort(([aStart, aEnd], [bStart, bEnd]) =>
      aEnd !== bEnd ? aEnd - bEnd : aStart - bStart
    );

    return this.getGaps(intervals);
  }

  getGaps = (intervals, gaps = 1) => {
    const prev = intervals.shift();

    for (const curr of intervals) {
      const [prevStart, prevEnd] = prev;
      const [currStart, currEnd] = curr;

      const hasGap = prevEnd <= currStart;
      if (!hasGap) continue;

      prev[1] = curr[1];
      gaps++;
    }

    return intervals.length + 1 - gaps;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    [1, 2],
    [2, 3],
    [3, 4],
    [1, 3],
  ],
  [
    [1, 2],
    [1, 2],
    [1, 2],
  ],
  [
    [1, 2],
    [2, 3],
  ],
];
test.quantify(testCases);
